File: bindings/java/samples/Zlink.Samples/src/main/java/dev/kairoscode/zlink/samples/MonitorRecvSample.java
package dev.kairoscode.zlink.samples;

import dev.kairoscode.zlink.Context;
import dev.kairoscode.zlink.MonitorEventType;
import dev.kairoscode.zlink.PairSocket;

public final class MonitorRecvSample {
    public static void main(String[] args) {
        SampleSupport.ensureNative();
        String endpoint = SampleSupport.tcpEndpoint();

        try (Context ctx = new Context();
             PairSocket server = new PairSocket(ctx);
             PairSocket client = new PairSocket(ctx);
             var serverMonitor = server.monitorOpen(
                 dev.kairoscode.zlink.MonitorEventType.CONNECTION_READY);
             var clientMonitor = client.monitorOpen(
                 dev.kairoscode.zlink.MonitorEventType.CONNECTION_READY)) {
            server.bind(endpoint);
            client.connect(endpoint);

            var serverEvent = serverMonitor.recv();
            var clientEvent = clientMonitor.recv();
            if (serverEvent.event() != MonitorEventType.CONNECTION_READY
                || clientEvent.event() != MonitorEventType.CONNECTION_READY) {
                throw new IllegalStateException("expected connection-ready events");
            }
            System.out.println("[monitor/recv] recv: \"connection-ready\"");
        }
    }
}

File: bindings/cpp/samples/monitor_recv_sample.cpp
/* SPDX-License-Identifier: MPL-2.0 */

#include "sample_common.hpp"

int main ()
{
    zlink::context_t ctx;
    zlink::pair_socket_t server (ctx);
    zlink::pair_socket_t client (ctx);
    zlink::monitor_handle_t server_monitor =
      server.monitor_handle (zlink::monitor_event::connection_ready);
    zlink::monitor_handle_t client_monitor =
      client.monitor_handle (zlink::monitor_event::connection_ready);

    assert (!server_monitor.recv (zlink::non_blocking_t {}));
    assert (!client_monitor.recv (zlink::non_blocking_t {}));

    assert (server.bind ("tcp://127.0.0.1:0") == 0);
    const std::string endpoint = server.options ().last_endpoint ();
    assert (!endpoint.empty ());
    assert (client.connect (endpoint) == 0);

    const zlink::monitor_event_t server_event = server_monitor.recv ();
    const zlink::monitor_event_t client_event = client_monitor.recv ();
    assert (server_event.event == zlink::monitor_event::connection_ready);
    assert (client_event.event == zlink::monitor_event::connection_ready);
    assert (!server_monitor.recv (zlink::non_blocking_t {}));
    assert (!client_monitor.recv (zlink::non_blocking_t {}));

    std::printf (
      "[monitor/recv] recv: \"connection-ready\" -> recv(non_blocking): empty\n");
    return 0;
}
